### PR TITLE
[WIP] Refactor the Lokomotive cluster configuration

### PR DESCRIPTION
**Pull request by ipochi**
This PR mostly deals with refactoring on two ends targeting issue #123 whilst also adding unit tests for the code base.

(a) How the end user creates configuration
(b) Organizing the code base 

(a) During discussion with @invidian @joaquimrocha  it was found that the current way of creating the cluster configuration is mixed with different types of fields and it could be much more cleaner, also the configuration is very inflexible, meaning fields which are not platform specific should be removed from the `cluster` block and separated to be able to put them in separate `.lokocfg` files.

This PR introduces  a new way of creating Lokomotive configuration.
```
controller {
  count = 3
  ssh_pubkeys = ["ssh-key"]
}

network {
  pod_cidr = "10.2.0.0/16"
  service_cidr = "10.3.0.0/16"
}

flatcar {
  version = "current"
  channel = "stable"
}

metadata {

}

cluster {
  // general cluster related fields
  cluster_domain_suffix  = "cluster.local"
}

platform "placket/aws/baremetal" {
  dns {
  ...
  }
  flatcar {
    arch = "amd64"
  }
  network {
    management_cidr = ["0.0.0.0/16"]
    node_private_cidr = "10.0.0.0/16"
  }
  auth_token = "packet_token"
  facility = "ams1"
  project_id = "project-id"
}
```

Top level blocks (flatcar, metadata, controller, network, cluster,platform) can be separated out into different lokocfg files for resuability.

`platform` block is platform specific.

(b) In order to achieve (a) there had to be done some code organization

Mainly this is done by introducing three main go structs `HCLLoader`, `LokomotiveConfig` and `Lokomotive`.

HCLLoader loads and parses the user provided hcl configuration in LokomotiveConfig
LokomotiveConfig deals with management of the configuration.
Lokomotive deals with the cluster operations

A new package `cluster` is introduced which contains the above struct 

Lokomotive in `pkg/cluster`
LokomotiveConfig in `pkg/cluster/config`

All cluster related operations in `cmd` package are consolidated in `pkg/cluster.

Note: I've tried to break the commits as much as I can , hence they are not atomic in nature.

### CODE REVIEW

File: pkg/components/external-dns/component_test.go [hunk 1]
```diff
@@ -61,112 +56,8 @@ func TestDefaultValues(t *testing.T) {
...
-   policy = "upsert-only"
-   owner_id = "test-owner"
-   aws {
-     zone_id = "TESTZONEID"
-     zone_type = "public"
-   }
- }
- `
-	// Unset AWS environment variables
-	os.Unsetenv("AWS_ACCESS_KEY_ID")
```
Comment (invidian) [inline pkg/components/external-dns/component_test.go:79]: Maybe `os.Setenv("key", "")` would make tests pass?

File: pkg/config/config.go [hunk 2]
```diff
@@ -159,8 +220,8 @@ func LoadConfig(lokocfgPath, lokocfgVarsPath string) (*Config, hcl.Diagnostics)
 	}
 
 	return &Config{
```
Comment (invidian) [inline pkg/config/config.go:222]: This struct should be called `HCLConfig`.

Comment (invidian) [reply pkg/config/config.go:222]: Maybe it would make sense to move it all to `pkg/config/hcl` then?

File: pkg/config/hcl_loader.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	hclConfig, diags := ParseHCLFiles(configFiles, variablesFile)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	return parseToLokomotiveConfig(hclConfig)
+}
+
+// parseToLokomotiveConfig parses the Config instance to LokomotiveConfig
+func parseToLokomotiveConfig(hclConfig *Config) (*config.LokomotiveConfig, hcl.Diagnostics) {
```
Comment (invidian) [inline pkg/config/hcl_loader.go:52]: Ditto HCL prefix. Or we should make it a method of `Config` struct (which should be renamed to `HCLConfig`.

File: pkg/config/hcl_loader.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	hclConfig, diags := ParseHCLFiles(configFiles, variablesFile)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	return parseToLokomotiveConfig(hclConfig)
+}
+
+// parseToLokomotiveConfig parses the Config instance to LokomotiveConfig
```
Comment (invidian) [inline pkg/config/hcl_loader.go:51]: All comments should end with `.`.
```suggestion
// parseToLokomotiveConfig parses the Config instance to LokomotiveConfig.
```

File: pkg/platform/aws/aws.go [hunk 5]
```diff
@@ -47,126 +43,99 @@ type workerPool struct {
...
-	NetworkMTU               int               `hcl:"network_mtu,optional"`
-	HostCIDR                 string            `hcl:"host_cidr,optional"`
-	PodCIDR                  string            `hcl:"pod_cidr,optional"`
-	ServiceCIDR              string            `hcl:"service_cidr,optional"`
-	ClusterDomainSuffix      string            `hcl:"cluster_domain_suffix,optional"`
-	EnableReporting          bool              `hcl:"enable_reporting,optional"`
-	CertsValidityPeriodHours int               `hcl:"certs_validity_period_hours,optional"`
-	WorkerPools              []workerPool      `hcl:"worker_pool,block"`
+type flatcar struct {
+	OSName string `hcl:"os_name,optional"`
```
Comment (invidian) [inline pkg/platform/aws/aws.go:47]: CoreOS is deprecated, I think we could remove it completely.

File: pkg/platform/aws/aws.go [hunk 6]
```diff
@@ -47,126 +43,99 @@ type workerPool struct {
...
 }
 
-// init registers aws as a platform
-func init() {
-	platform.Register("aws", NewConfig())
+type network struct {
+	HostCIDR string `hcl:"host_cidr"`
```
Comment (invidian) [inline pkg/platform/aws/aws.go:51]: In a way, we use the same in Packet, except here we use it for setting the CIDR and for Packet to set the firewall. I think that should be moved to common fields.

File: pkg/platform/aws/aws.go [hunk 7]
```diff
@@ -47,126 +43,99 @@ type workerPool struct {
...
-	if configBody == nil {
-		return hcl.Diagnostics{}
-	}
-
-	if diags := gohcl.DecodeBody(*configBody, evalContext, c); len(diags) != 0 {
-		return diags
-	}
+type controller struct {
+	Type        string   `hcl:"type,optional"`
+	CLCSnippets []string `hcl:"clc_snippets,optional"`
```
Comment (invidian) [inline pkg/platform/aws/aws.go:56]: Aren't CLC snippets supported on all platforms?

File: pkg/platform/aws/aws.go [hunk 8]
```diff
@@ -47,126 +43,99 @@ type workerPool struct {
...
-	}
+type config struct {
+	Metadata        *configpkg.Metadata
+	DNSZone         string       `hcl:"dns_zone"`
+	DNSZoneID       string       `hcl:"dns_zone_id"`
+	ExposeNodePorts bool         `hcl:"expose_nodeports,optional"`
+	CredsPath       string       `hcl:"creds_path,optional"`
+	Region          string       `hcl:"region,optional"`
+	WorkerPools     []workerPool `hcl:"worker_pool,block"`
+	Disk            *disk        `hcl:"disk,block"`
```
Comment (invidian) [inline pkg/platform/aws/aws.go:73]: It seems we should move controllers config into separate block, which would also use `workerPool` struct underneath.

File: pkg/platform/aws/aws.go [hunk 9]
```diff
@@ -228,10 +200,86 @@ func (c *config) checkValidConfig() hcl.Diagnostics {
...
 }
 
+func validCIDR(cidr string) error {
+	_, _, err := net.ParseCIDR(cidr)
+
+	return err
+}
+func (c *config) checkFlatcarConfig() hcl.Diagnostics {
+	var diagnostics hcl.Diagnostics
+	if c.Flatcar.OSName != "flatcar" {
```
Comment (invidian) [inline pkg/platform/aws/aws.go:259]: We should definitely remove this field :)

File: pkg/platform/baremetal/baremetal.go [hunk 10]
```diff
@@ -16,186 +16,137 @@ package baremetal
...
-	OSChannel              string   `hcl:"os_channel,optional"`
-	OSVersion              string   `hcl:"os_version,optional"`
-	SSHPubKeys             []string `hcl:"ssh_pubkeys"`
-	WorkerNames            []string `hcl:"worker_names"`
-	WorkerMacs             []string `hcl:"worker_macs"`
-	WorkerDomains          []string `hcl:"worker_domains"`
-}
+type controller struct {
+	Domains []string `hcl:"controller_domains"`
+	Macs    []string `hcl:"controller_macs"`
```
Comment (invidian) [inline pkg/platform/baremetal/baremetal.go:29]: Should be `MACs` I think.

File: pkg/platform/packet/packet.go [hunk 11]
```diff
@@ -50,161 +48,117 @@ type workerPool struct {
...
+	Arch          string `hcl:"os_arch,optional"`
+	IPXEScriptURL string `hcl:"ipxe_script_url,optional"`
 }
 
-// init registers packet as a platform
-func init() {
-	platform.Register("packet", NewConfig())
+type network struct {
+	ManagementCIDRs []string `hcl:"management_cidrs"`
+	NodePrivateCIDR string   `hcl:"node_private_cidr"`
```
Comment (invidian) [inline pkg/platform/packet/packet.go:58]: This is kind of the same as `HostCIDR` in AWS.

File: pkg/terraform/executor.go [hunk 12]
```diff
@@ -96,8 +96,23 @@ type Executor struct {
 	logger        *log.Entry
 }
 
-// NewExecutor initializes a new Executor.
-func NewExecutor(conf Config) (*Executor, error) {
+// InitializeExecutor initialized Terraform directory using given backend and platform
+// and returns configured executor.
+func InitializeExecutor(assetDir string, verbose bool) (*Executor, error) {
```
Comment (invidian) [inline pkg/terraform/executor.go:101]: This helper seems weird. I think Terraform package should have no knowledge about "asset dir".

**Comment by rata** [comment]
@ipochi thanks a lot for the PR. It seems like a lot of work! =)

I haven't looked at the code at all. But a meta comment:
* Hopefully is not the case now, but I think others can spot flaws or things to change about any proposed big change, and is good to have the input of others. However, doing so when the code is already written for some design, is not good as is way more expensive and painful to change (or we end-up don't doing them because it is too expensive). If, after some design is proposed, others can have a look at share their input, we can easily improve the design before writing code that maybe goes in another direction. I don't want to add bureaucracy, but I think a very thin doc for team-approval might help. What do others think of this for the future?

Or maybe we can move the discussion to another place. But wanted to raise the issue here, though.

I guess this next questions are for people in the meeting cc @joaquimrocha @ipochi @invidian 

On one hand, shouldn't we make these changes with multi-cluster support in mind? Is this taking into account that we will want to go the multi-cluster route in the future?

On the other hand, leaving that aside, I just read the PR description and I really don't understand why some fields are split. Why is flatcar (and others) specified at two different levels (root and inside platform)? I don't follow at all. In particular, what is the reasoning for Flatcar being split?

Why a top-root Flatcar block with version and channel and then the architecture is on the platform scoped flatcar block? How is the Flatcar root block used (or not) inside the platform? Will all the nodes use the "union" config of all flatcar fields around (like flatcar root and flatcar inside platform, didn't check if it can be specified in other blocks)? This is not clear, from a user POV, IMHO.

Or looking at others, what is the reasoning to know where to look for something (like `pod_cidr`, or other fields)? For me, it is not intuitive to know where to go and look for an option. Here is an example, in case it is not clear, of some things that are not clear or don't seem consistent to me:
*  if you want to change the controller, go to root `controller` block. But if you want to change a worker pool, look inside the platform block. Why such discrepancy? (maybe I'm wrong here, just tried to look at the code to see where worker pools falls)
* Want to change the `pod_cidr` used in the cluster? then go to `network` at root (this is okay, but confusing when doing the reading the next two items)
* want to change the DNS? go to platform (I really don't follow this!)
* want to change the cluster suffix used in DNS names? go to `cluster` root block (not platform nor  root network block, as previous examples might make you think).

I honestly don't follow the reasoning, really. It just seems inconsistent to me :-/

I'm sure I'm missing something, can you please share the reasoning to split things? So is obvious where to go for when you want to change something. It will be good if that can be documented so it is clear to users too.

Also, IIUC, does this mean that with this change I can't have some worker pool running Flatcar beta, then? Or is it possible because worker pools are specified inside platform block? I guess, then, the "flatcar" root block applies to controllers? Or maybe is it used "by default" when worker pools don't specify nothing about flatcar? It's really confusing for me :)

Another concern that I have is that it seems we will have longer config files (to express the same as before). Hopefully they will be more clear, but... I don't know, I'm missing the reasoning on why the filed are split :)

**Comment by invidian** [comment]
Hey @rata, thanks for looking into this PR. I'll try to answer your concerns.

>Hopefully is not the case now, but I think others can spot flaws or things to change about any proposed big change, and is good to have the input of others. However, doing so when the code is already written for some design, is not good as is way more expensive and painful to change (or we end-up don't doing them because it is too expensive). If, after some design is proposed, others can have a look at share their input, we can easily improve the design before writing code that maybe goes in another direction. I don't want to add bureaucracy, but I think a very thin doc for team-approval might help. What do others think of this for the future?

CC @iaguis 

I agree, that for bigger changes we could have some team approval process before actual work happens. However, in this particular case, I have a feeling, that it wouldn't really work, because of complexity of the change. IMO it is easier to iterate on already working code rather than spend a lot of time on the doc describing those changes (as it may be misaligned a lot with the reality later on). 

>On one hand, shouldn't we make these changes with multi-cluster support in mind? Is this taking into account that we will want to go the multi-cluster route in the future?

I'm not sure about this one. For me, it seems that this PR is actually not that bad step for multi-cluster supports, as the only thing we would have to do later on, is to wrap all the structs into another `Cluster` one. Not sure how this could be represented in HCL in clear way though.

>On the other hand, leaving that aside, I just read the PR description and I really don't understand why some fields are split. Why is flatcar (and others) specified at two different levels (root and inside platform)? I don't follow at all. In particular, what is the reasoning for Flatcar being split?

The idea here is to keep configuration options which are common to all platforms in one place and be able to set platform-specific values in `platform` block. This way, when changing the platform for your cluster, you only need to "translate" one block (`platform`).

>want to change the DNS? go to platform (I really don't follow this!)

IIUC, there is still pending work on moving all fields into the right place and this one has not been moved yet. This is to limit the amount of work which needs to be done. I was suggesting @ipochi to just move one field for the initial iteration of this work, to reduce the confusion ;)

>want to change the cluster suffix used in DNS names? go to cluster root block (not platform nor root network block, as previous examples might make you think).

Well, naming things is hard :smile: that's why we need reviewers to find a good place, where you would expect them to be for all fields we support.

>Also, IIUC, does this mean that with this change I can't have some worker pool running Flatcar beta, then? Or is it possible because worker pools are specified inside platform block? I guess, then, the "flatcar" root block applies to controllers? Or maybe is it used "by default" when worker pools don't specify nothing about flatcar? It's really confusing for me :)

That seems like a regression, which shouldn't happen. I think we could have top level `flatcar` block and per worker/controller block as well for the overrides.

**Comment by surajssd** [comment]
- I agree with Rodrigo that a design discussion is due, we need a way to know what this refactor is bringing in. A doc would be very helpful. And first we should discuss the organisation of the config and only then start working on it in the code.

- The use of word refactor is kinda misleading here, because apart from refactor of the code in the background there are massive changes to the frontend that are concerned with the user. Backwards incompatible changes. So a design doc is necessary.

- Since we are bringing in these radical changes, we might as well think of moving to plain yaml. Personally I am not a huge fan of HCL. And I distantly remember we had some discussion around if might want to consider moving to yaml, replacing HCL.

- Having design discussion and PR reviews mixed on the same page makes it harder for everyone. The code author is inundated in the comments and someone looking at the changes for the first time gives up because it is incomprehensible amount of information. Also github gives you (and your browser) hard time after certain number of comments.

---

I would suggest that we start with design discussion, get it finalised; then nobody will have objection on the PR once implementation is going on.

But if author of the work (in this case Imran) wants to do implementation on the sides to clarify the nuances in the design they are most welcome to do so. In such case the PR should be exempted from code reviews because it is implementing a highly changing design proposal.

**Comment by surajssd** [comment]
I reserve my design related comments until the design doc is ready.

**Comment by iaguis** [comment]
It seems to me a thin design doc explaining the user-facing changes would help since it's important to have buy-in from everybody in such a big user-affecting change and we have some good points raised above.

I think the work invested in this PR is very useful to gain insights on what the implementation could look like and what's possible within reason given the current state of the code.

So what about @ipochi writing a small design doc with the config reorganization so others can comment? From what I understand moving things around shouldn't be too difficult with this new structure and as I said I trust @ipochi has some good insights on what's possible and can provide good feedback there.

**Comment by ipochi** [comment]
@rata @invidian @surajssd @iaguis 

Yes I do, believe a design document is needed and I will provide one to ease concerns.

Also, I'd like to point out that the refactoring design decision was never done in isolation (myself, @invidian and @joaquimrocha  were involved in the approach). Also an outsiders perspective in the form of @joaquimrocha was very helpful in formulating the structure.

The basic idea is that in 
 - non-platform specific fields shouldn't have to be re declared and organized properly.
 - the flexibility of all top level blocks to be able to reside in their own `*.lokocfg` files if needed. 

Regarding the naming of the blocks , I took the liberty of naming them which in hindsight I shouldn't have done so. however changing the block names would be minor effort.

@surajssd I also believe that, we shouldn't be too much concerned about backward compatibility issues as we are in 0.1.0 release. Not making breaking changes now  for the sake of being backwards compatible will set us back in terms of what Lokomotive can do and how easy it could become for the user to write configurations.

More releases down the line, the it would become harder and harder to introduce backwards compatibility changes.

@rata  I have not touched upon the way to define worker pools, so there is some redundancy there. The reason I did not touch upon that is the scope of the PR was getting bigger and bigger.

**Comment by ipochi** [comment]
@invidian @iaguis @rata @surajssd @joaquimrocha @johananl 

Design document - https://docs.google.com/document/d/1oPMdb02pf1guDdmLv0dxLIJ1ugh21odJw15o1PMKFp4/edit?usp=sharing

**Comment by ipochi** [comment]
closing this.

**Pull request closed without merging**